cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -524,21 +529,45 @@ public RunId runInternal(ProgramId programId, Map<String, String> userArgs, Map<
     userId = userId == null ? "" : userId;
 
     checkCapability(programDescriptor);
-    synchronized (this) {
-      if (maxConcurrentRuns > 0 && maxConcurrentRuns <= store.countActiveRuns(maxConcurrentRuns)) {
-        ConflictException e = new ConflictException(
-          String.format("Program %s cannot start because the maximum of %d concurrent runs is exceeded",
-                        programId, maxConcurrentRuns));
 
-        programStateWriter.reject(programId.run(runId), programOptions, programDescriptor, userId, e);
-        throw e;
-      }
+    ProgramRunId programRunId = programId.run(runId);
+    RunRecordCounter.Count count = runRecordCounter.addRequestAndGetCount(programRunId);
+
+    if (maxConcurrentRuns >= 0 && maxConcurrentRuns < count.getRunsCount()) {
+      TooManyRequestsException e = new TooManyRequestsException(
+        String.format("Program %s cannot start because the maximum of %d concurrent running runs is allowed",
+                      programId, maxConcurrentRuns));
+
+      runRecordCounter.removeRequest(programRunId);
+      programStateWriter.reject(programRunId, programOptions, programDescriptor, userId, e);
+      throw e;
     }
 
-    LOG.info("Attempt to run {} program {} as user {} with arguments {}", programId.getType(), programId.getProgram(),
-             authenticationContext.getPrincipal().getName(), userArgs);
+    if (maxConcurrentLaunching >= 0 && maxConcurrentLaunching < count.getLaunchingCount()) {
+      TooManyRequestsException e = new TooManyRequestsException(
+        String.format("Program %s cannot start because the maximum of %d concurrent provisioning/starting runs" +
+                        " is allowed", programId, maxConcurrentLaunching));
+
+      runRecordCounter.removeRequest(programRunId);
+      programStateWriter.reject(programRunId, programOptions, programDescriptor, userId, e);
+      throw e;
+    }
+
+      LOG.info("Attempt to run {} program {} as user {} with arguments {}", programId.getType(), programId.getProgram(),
+               authenticationContext.getPrincipal().getName(), userArgs);
+
+    boolean done = false;
+    try {
```

Comment:
Could you put the whole thing starting from line 536 into this try? This will be more safe and save you few explicit `removeRequest` calls

Reply:
Done!